substrate/client/network/sync/src/strategy/state.rs

```diff
@@ -0,0 +1,758 @@
+		let peer_id = PeerId::random();
+		let initial_peers = std::iter::once((peer_id, 10));
+		let mut state_strategy =
+			StateStrategy::new_with_provider(Box::new(state_sync_provider), initial_peers);
+		// Manually set the peer's state .
+		state_strategy.peers.get_mut(&peer_id).unwrap().state = PeerState::DownloadingState;
+
+		let dummy_response = OpaqueStateResponse(Box::new(StateResponse::default()));
+		state_strategy.on_state_response(peer_id, dummy_response);
+
+		// No actions generated.
+		assert_eq!(state_strategy.actions.len(), 0)
+	}
+
+	#[test]
+	fn complete_state_response_leads_to_block_import() {
+		// Build block to use for checks.
+		let client = Arc::new(TestClientBuilder::new().set_no_genesis().build());
+		let mut block_builder = BlockBuilderBuilder::new(&*client)
+			.on_parent_block(client.chain_info().best_hash)
+			.with_parent_block_number(client.chain_info().best_number)
+			.build()
+			.unwrap();
+		block_builder.push_storage_change(vec![1, 2, 3], Some(vec![4, 5, 6])).unwrap();
+		let block = block_builder.build().unwrap().block;
+		let header = block.header().clone();
+		let hash = header.hash();
+		let body = Some(block.extrinsics().iter().cloned().collect::<Vec<_>>());
+		let state = ImportedState { block: hash, state: KeyValueStates(Vec::new()) };
+		let justifications = Some(Justifications::from((*b"FRNK", Vec::new())));
+
+		// Prepare `StateSync`
+		let mut state_sync_provider = MockStateSync::<Block>::new();
+		let import = ImportResult::Import(
+			hash,
+			header.clone(),
+			state.clone(),
+			body.clone(),
+			justifications.clone(),
+		);
+		state_sync_provider.expect_import().return_once(move |_| import);
+
+		// Reference values to check against.
+		let expected_origin = BlockOrigin::NetworkInitialSync;
+		let expected_block = IncomingBlock {
+			hash,
+			header: Some(header),
+			body,
+			indexed_body: None,
+			justifications,
+			origin: None,
+			allow_missing_state: true,
+			import_existing: true,
+			skip_execution: true,
+			state: Some(state),
+		};
+		let expected_blocks = vec![expected_block];
+
+		// Prepare `StateStrategy`.
+		let peer_id = PeerId::random();
+		let initial_peers = std::iter::once((peer_id, 10));
+		let mut state_strategy =
+			StateStrategy::new_with_provider(Box::new(state_sync_provider), initial_peers);
+		// Manually set the peer's state .
+		state_strategy.peers.get_mut(&peer_id).unwrap().state = PeerState::DownloadingState;
+
+		// Receive response.
+		let dummy_response = OpaqueStateResponse(Box::new(StateResponse::default()));
+		state_strategy.on_state_response(peer_id, dummy_response);
+
+		assert_eq!(state_strategy.actions.len(), 1);
+		assert!(matches!(
+			&state_strategy.actions[0],
+			StateStrategyAction::ImportBlocks { origin, blocks }
+				if *origin == expected_origin && *blocks == expected_blocks,
+		));
+	}
+
+	#[test]
+	fn importing_unknown_block_doesnt_finish_strategy() {
+		let target_hash = Hash::random();
```

Comment:
Maybe you could used fixed value here?
Just in case logs from different executions need to be compared?
Same applies to other random values.

Reply:
Yeah, I also don't really like the undetermined random numbers in tests. But constructing predefined values, especially when we init 10 peers, would be real PITA.